packages/teleterm/src/ui/Navigator/ExpanderConnections/useExpanderConnections.ts

```diff
@@ -18,81 +18,108 @@ import * as Icons from 'design/Icon';
   const ctx = useAppContext();
-  const items = initItems(ctx);
+  const items = getWorkspaceItems(ctx);
+  const [, rerender] = React.useState<any>();
 
   // subscribe to services
   ctx.clustersService.useState();
   ctx.docsService.useState();
 
+
+  //TODO: think how to use useStore for services from context bridge
+  useEffect(() => {
+    function onChange() {
+      rerender({});
+    }
+
+    ctx.workspaceService.subscribe(onChange);
+    return () => ctx.workspaceService.unsubscribe(onChange);
+  }, []);
+
   return {
-    serviceClusters: ctx.clustersService,
     items,
+    onItemOpen: (item: ConnectionItem) => handleItemOpen(ctx, item),
+    onItemRemove: (item: ConnectionItem) => handleItemRemove(ctx, item),
   };
 }
 
-function initItems(ctx: AppContext): ConnectionItem[] {
-  const gateways = ctx.clustersService.getGateways();
-  const tshDocs = ctx.docsService.getTshNodeDocuments();
-
-  const demoHistoryItems: ConnectionItem[] = [
-    {
-      title: 'root@node1',
-      kind: 'nav.connection-tsh',
-      uri: '/nothing/402',
-      Icon: Icons.Keypair,
-      items: [],
-      group: false,
-      status: 'disconnected',
+function getWorkspaceItems(ctx: AppContext): ConnectionItem[] {
+  return [
+    ...ctx.workspaceService
+      .get()
+      .recentDocuments.map(document => {
+        if (document.kind === 'doc.terminal_tsh_node') {
+          return { ...document, status: 'disconnected' as const };
+        } else {
+          return document;
+        }
+      })
+      .map(document => createConnectionItem(ctx, document)),
+  ].sort(item => {
+    if (item.status === 'connected') {
+      return -1;
+    }
+    return 1;
+  });
+}
+
+function handleItemOpen(ctx: AppContext, item: ConnectionItem) {
+  if (!ctx.docsService.getDocuments().find(d => d.uri === item.uri)) {
+    ctx.docsService.add(item.document);
+  }
+  ctx.docsService.open(item.uri);
+}
+
+function handleItemRemove(ctx: AppContext, item: ConnectionItem) {
+  const recentDocuments = ctx.workspaceService
+    .get()
+    .recentDocuments.filter(d => d.uri !== item.uri);
+  ctx.workspaceService.update({ recentDocuments });
+}
+
+function createConnectionItem(
+  ctx: AppContext,
+  document: Document
+): ConnectionItem {
+  function getStatus() {
+    switch (document.kind) {
+      case 'doc.terminal_tsh_node':
+        return (ctx.docsService.getDocument(document.uri) as DocumentTshNode)
+          ?.status === 'connected'
+          ? 'connected'
+          : 'disconnected';
+      case 'doc.gateway':
+        return ctx.clustersService.findGateway(document.uri)
+          ? 'connected'
+          : 'disconnected';
+    }
+  }
+
+  return {
+    get status() {
+      return getStatus();
     },
-    {
-      title: 'pearl@node1',
-      kind: 'nav.connection-tsh',
-      uri: '/nothing/2cf',
-      Icon: Icons.Keypair,
-      items: [],
-      group: false,
-      status: 'disconnected',
+    get uri() {
+      return document.uri;
```

Comment:
Lets keep simple objects without property accessors as I am not sure how it would behave with spread operators and such.